### ISSUE TITLE
Panic in coloring logic when using nested groups in 'filter' mode

### ISSUE THREAD

**Issue by obaudys** [zix99/rare#24]
First of all, let me say, this is an awesome project.  Nice work!  I was in the process of writing something similar but much much worse;  I think I might use this as a library instead! 

I've manage to cause a panic in the following scenario:
```
echo '1,2,3,4,5,6,7,8,9,0' | rare filter -m "(^[^,]*)(,([^,]*)){5}"
panic: runtime error: slice bounds out of range [11:10]

goroutine 1 [running]:
rare/pkg/color.WrapIndices(0xc000500000, 0x13, 0xc0000b4590, 0x6, 0x6, 0x206860, 0x1207860)
	/Users/ondrejb/Documents/git/rare/pkg/color/coloring.go:95 +0x832
rare/cmd.filterFunction(0xc00024c840, 0x0, 0xc0004752d0)
	/Users/ondrejb/Documents/git/rare/cmd/filter.go:33 +0x276
github.com/urfave/cli.HandleAction(0x13b20c0, 0x144c828, 0xc00024c840, 0xc00024c840, 0x0)
	/Users/ondrejb/go/pkg/mod/github.com/urfave/cli@v1.22.2/app.go:523 +0xfd
github.com/urfave/cli.Command.Run(0x142c0a8, 0x6, 0x1429f04, 0x1, 0x0, 0x0, 0x0, 0x1441dfa, 0x44, 0x0, ...)
	/Users/ondrejb/go/pkg/mod/github.com/urfave/cli@v1.22.2/command.go:174 +0x58e
github.com/urfave/cli.(*App).Run(0xc0004d6000, 0xc000090040, 0x4, 0x4, 0x0, 0x0)
	/Users/ondrejb/go/pkg/mod/github.com/urfave/cli@v1.22.2/app.go:276 +0x7d4
main.cliMain(0xc000090040, 0x4, 0x4, 0x0, 0x0)
	/Users/ondrejb/Documents/git/rare/main.go:101 +0x666
main.main()
	/Users/ondrejb/Documents/git/rare/main.go:105 +0x49
```
To save you parsing manually,  the three groups here are the 1st column of a csv, the 6th column but including the leading comma, and the 6th field without the leading comma.

When I make the second (outer) group non capturing, ie.  `(^[^,]*)(?:,([^,]*)){5}` everything works fine and I get the 1st and 6th field (groups {1} and {3}).   Obviously, if I use the --nocolor option, or if I use an expression eg. `-e '{1} {2} {3}`, everything is fine.

I haven't looked deep into the code yet, but obviously since the match groups overlap, the starting index of the inner group lies inside the outer group, and the colouring logic doesn't account for this scenario.

I'd suggest that the inner match should take precedence when colouring the matching text (ie.  inner match colours "overwrite" the outer group)

I'll have a crack at making a pull request to fix this myself soon.

**Comment by zix99** [comment]
Thanks for filing this, and I'm glad you enjoy the project!  Feel free to take a stab at it, and ping me, and I'll give it a look.

**Comment by obaudys** [comment]
I've added a trivial fix to stop the panic with #25, as well as some test cases (looks like the code coverage test is failing but that's not my fault).

Actually getting it to resume colouring the outer group is pretty tricky,  I've got some messy code but it doesn't reset the colour back.  eg missing yellow here:

![image](https://user-images.githubusercontent.com/53287664/96069193-67687580-0ee9-11eb-8934-f5ee52c7edcf.png)

Here's the code to do that, I can't think of a good way to go back to the yellow in the example above:

```diff --git a/pkg/color/coloring.go b/pkg/color/coloring.go
index d6ad1d1..a0c5add 100644
--- a/pkg/color/coloring.go
+++ b/pkg/color/coloring.go
@@ -85,12 +85,36 @@ func WrapIndices(s string, groups []int) string {
 
        var sb strings.Builder
        lastIndex := 0
+       overrideColors := make(map[int]int, 0)
 
        for i := 0; i < len(groups); i += 2 {
                start := groups[i]
                end := groups[i+1]
+
+               j, ok := overrideColors[i]
+               if !ok {
+                       j = i
+               }
+               // check to see if the next group is an inner group
+               if i+3 < len(groups) {
+                       nextStart := groups[i+2]
+                       nextEnd := groups[i+3]
+
+                       if end > nextEnd {
+                               //  create an additional group to resume coloring
+                               groups = append(groups, 0, 0)
+                               copy(groups[i+5:], groups[i+3:])
+                               groups[i+4] = nextEnd
+                               groups[i+5] = end
+                               overrideColors[i+4] = j
+                       }
+                       if end > nextStart {
+                               // stop coloring before start of next group
+                               end = nextStart
+                       }
+               }
                if start >= 0 && end >= 0 {
-                       color := groupColors[(i/2)%len(groupColors)]
+                       color := groupColors[(j/2)%len(groupColors)]
 
                        sb.WriteString(s[lastIndex:start])
                        sb.WriteString(string(color))

```

**Comment by zix99** [comment]
I'll take a look and merge in the simple change if it fixes the panic.  Does it work for triple-nested groups (and n-nested groups) as well?  Looking at it briefly, it might not.  At a certain point there could be diminishing return, and maybe we jneed new checks to make sure the bounds are valid (eg `start >= 0 && end >= 0 && end > start && start >= lastIndex`.  Not perfect, and won't color correctly, but should prevent a panic).

Related to the code in this issue, I am slightly worried about adding a `map` to this function, as it represents a new allocation on a high-volume function that could significantly impact performance.  If that's the best way to solve the larger problem, please consider writing some benchmark's for this function.  I'm also not sure it addresses the triple-nesting/n-nesting concern.

Thoughts?

**Comment by obaudys** [comment]
It does indeed work for triple nested groups.  But your suggestion `if start >= 0 && end >= 0 && end > start && start >= lastIndex`  performs much better.    

I think it's better have good performance vs accurate colouring.  Plus to get the colouring right, I think of an algorithm that doesn't use allocations;  I'd thought perhaps to put the groups into a tree, along with the colour, and then split the nodes of the tree if a group falls inside another node.  That's needlessly expensive.

I'll update the pull request with your suggestion as well as a the test (which would fail on colour checking) and add a benchmark.

**Comment by zix99** [comment]
Resolved with #25
